### PR TITLE
Re-work JS trace filtering

### PR DESCRIPTION
**Pull request by armanbilge**
If this works, hopefully we don't ever have to think about this again.

### CODE REVIEW

File: core/shared/src/main/scala/cats/effect/tracing/Tracing.scala [hunk 1]
```diff
@@ -48,18 +56,29 @@ private[effect] object Tracing extends TracingPlatform {
...
     while (idx < len) {
       val methodSite = stackTrace(idx - 1)
       val callSite = stackTrace(idx)
-      val callSiteClassName = callSite.getClassName
-      val callSiteMethodName = callSite.getMethodName
-      val callSiteFileName = callSite.getFileName
 
-      if (!applyStackTraceFilter(callSiteClassName, callSiteMethodName, callSiteFileName))
+      if (isIOClass(methodSite.getClassName()) && isExternalClass(callSite.getClassName()))
         return combineOpAndCallSite(methodSite, callSite)
```
Comment (armanbilge) [inline core/shared/src/main/scala/cats/effect/tracing/Tracing.scala]: What we are looking for is actually quite simple: an `IO` method being called from some external code.

**Comment by armanbilge** [comment]
https://typelevel.org/cats-effect/docs/tracing#complete-code
```
The 20th fibonacci number is 17711
[error] (run-main-3) java.lang.Throwable: Boom!
[error] java.lang.Throwable: Boom!
[error]         at Example$.$anonfun$program$5(Example.scala:36)
[error]         at apply @ Example$.$anonfun$program$3(Example.scala:35)
[error]         at ifM @ Example$.$anonfun$program$3(Example.scala:36)
[error]         at apply @ Example$.$anonfun$program$1(Example.scala:34)
[error]         at flatMap @ Example$.$anonfun$program$1(Example.scala:34)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at apply @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.fib(Example.scala:24)
[error]         at flatMap @ Example$.program(Example.scala:33)
[error]         at run$ @ Example$.run(Example.scala:21)
```

File: core/shared/src/main/scala/cats/effect/tracing/Tracing.scala (hunk 1 above)
Comment (armanbilge) [reply core/shared/src/main/scala/cats/effect/tracing/Tracing.scala]: Lol nope 🙃 not that simple anyway. The method called doesn't have to have anything to do with `IO`. My bad.

**Comment by armanbilge** [comment]
```
[error] cats.effect.IOFiber@6fbe93f9 WAITING
[error]  ├ flatMap @ Deadlock$.$anonfun$run$2(Example.scala:26)
[error]  ├ flatMap @ Deadlock$.$anonfun$run$1(Example.scala:25)
[error]  ├ deferred @ Deadlock$.<clinit>(Example.scala:22)
[error]  ├ flatMap @ Deadlock$.<clinit>(Example.scala:22)
[error]  ╰ run$ @ Deadlock$.run(Example.scala:19)
[error]  
[error] cats.effect.IOFiber@5ba3776d WAITING
[error]  ├ get @ Deadlock$.$anonfun$run$1(Example.scala:24)
[error]  ╰ get @ Deadlock$.$anonfun$run$1(Example.scala:24)
[error]  
[error] Thread[io-compute-2,5,main] (#2): 0 enqueued
[error] Thread[io-compute-4,5,main] (#4): 0 enqueued
[error] Thread[io-compute-5,5,main] (#5): 0 enqueued
[error] Thread[io-compute-0,5,main] (#0): 0 enqueued
[error] Thread[io-compute-1,5,main] (#1): 0 enqueued
[error] Thread[io-compute-3,5,main] (#3): 0 enqueued
[error]  
[error] Global: enqueued 0, foreign 0, waiting 2
```

**Comment by armanbilge** [comment]
Ok, so this is _very_ close. Seems to be working on JVM.

For JS, the real issue is now illuminated: lambdas are encoded as `<jscode>.{anonymous}`, yes that is the class name and the method. So actually it is _impossible_ to know whether a lambda originated in internal or external code without a source map.

**Comment by armanbilge** [comment]
```
The 20th fibonacci number is 17711
java.lang.Throwable: Boom!
  at {anonymous}()(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:36:28)
  at ifM @ <jscode>.{anonymous}()(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:35)
  at flatMap @ <jscode>.{anonymous}()(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:34)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at delay @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/https:/raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937)
  at flatMap @ Example$.fib(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:24)
  at flatMap @ Example$.program(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:33)
  at main @ Example.main(/workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/file:/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:21)
```

**Comment by armanbilge** [comment]
Well now the sourcemaps seem to have gone completely bonkers 😕 

```
cats.effect.IOFiber@484de RUNNING: >> @ <jscode>.{anonymous}()(/home/runner/work/cats-effect/cats-effect/tests/js/target/scala-2.12/cats-effect-tests-test-fastopt/file:/home/runner/work/cats-effect/cats-effect/tests/shared/src/test/scala/cats/effect/kernel/SemaphoreSpec.scala:61)
```

Also having issues on my local (although this test passes strangely enough).

**Comment by armanbilge** [comment]
Slowly getting to the bottom of this (for a red herring see https://github.com/scala-js/scala-js/issues/4599).
```
cats.effect.tracing.TracingEvent$StackTrace
  at null.$thiz.Lcats_effect_tracing_TracingPlatform__f_calculateTracingEvent(https://raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/tracing/Tracing.scala:31:5)
  at scala.scalajs.runtime.AnonFunction1.apply(https://raw.githubusercontent.com/scala-js/scala-js/v1.7.1/library/src/main/scala/scala/scalajs/runtime/AnonFunctions.scala:27:38)
  at cats.effect.tracing.Tracing$.calculateTracingEvent(https://raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/js/src/main/scala/cats/effect/tracing/TracingPlatform.scala:39:26)
  at cats.effect.IO$.delay(https://raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:946:44)
* at Example$.fib(https://raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IO.scala:937:43)
  at null.<anonymous>(/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:26:12)
  at scala.scalajs.runtime.AnonFunction1.apply(https://raw.githubusercontent.com/scala-js/scala-js/v1.7.1/library/src/main/scala/scala/scalajs/runtime/AnonFunctions.scala:27:38)
  at null.$p_Lcats_effect_IOFiber__runLoop__Lcats_effect_IO__I__I__V(https://raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IOFiber.scala:408:22)
  at null.$p_Lcats_effect_IOFiber__execR__V(https://raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IOFiber.scala:1275:14)
  at cats.effect.IOFiber.run(https://raw.githubusercontent.com/typelevel/cats-effect/05f3eaae1e9ec00d43ffba36c16078bc1c1984e6/core/shared/src/main/scala/cats/effect/IOFiber.scala:124:22)
```

This is a raw tracing event trace for the fibonacci example above. I've *ed the line where things get weird: everything about that stackframe is correct, except it thinks its `Example$.fib` when in fact it is `IO$.apply`! Following the line number will show exactly that.

IDK whether this is specifically a bug, or it's because Scala.js is somehow eliding `apply` since it is after all only an alias for `delay`.

**Comment by armanbilge** [comment]
Update: it's definitely optimization. Re-defining `IO.apply` fixes the traces:
```scala
def apply[A](thunk: => A): IO[A] = {
  System.out.println("do something side-effecty")
  delay(thunk)
}
```

**Comment by armanbilge** [comment]
Helpful input from Seb on [Discord](https://discord.com/channels/632150470000902164/635668814956068864/913110451204788284):
> There's not much we can do about that. Source maps are too weak to accurately represent inlining. So that happens.
> If you particularly care about a specific method, mark it `@noinline`.

File: tests/shared/src/test/scala/cats/effect/IOFiberSpec.scala [hunk 2]
```diff
@@ -25,9 +25,9 @@ class IOFiberSpec extends BaseSpec {
 
   "IOFiber" should {
     "toString a running fiber" in real {
-      def loop: IO[Unit] = IO.unit >> loop
+      def loop: IO[Unit] = IO.unit.flatMap(_ => loop)
```
Comment (armanbilge) [inline tests/shared/src/test/scala/cats/effect/IOFiberSpec.scala:28]: Yeah, so aliases break JS tracing.

**Comment by armanbilge** [comment]
```
cats.effect.IOFiber@d WAITING
 
cats.effect.IOFiber@9 WAITING
 ╰ deferred @ <jscode>.null.$c_LDeadlock$(/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:22)
 
cats.effect.IOFiber@a WAITING
 ├ flatMap @ <jscode>.null.<anonymous>(/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:26)
 ├ flatMap @ <jscode>.null.<anonymous>(/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:25)
 ├ deferred @ <jscode>.null.$c_LDeadlock$(/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:22)
 ╰ flatMap @ <jscode>.null.$c_LDeadlock$(/workspace/cats-effect/example/js/src/main/scala/cats/effect/example/Example.scala:22)
 
cats.effect.IOFiber@b WAITING
 
Global: enqueued 0, waiting 4
```

**Comment by armanbilge** [comment]
So the fiber dump above is kind of disappointing. Why doesn't it indicate that the fiber is blocked on `get`?

Here's the raw tracing event stacktrace. I've indicated the `get` with `*`:
```
cats.effect.tracing.TracingEvent$StackTrace
  at null.<anonymous>(https://raw.githubusercontent.com/typelevel/cats-effect/f40765f2421487f83537dfda8829266970d6997c/core/shared/src/main/scala/cats/effect/tracing/Tracing.scala:31:5)
  at scala.scalajs.runtime.AnonFunction1.apply(https://raw.githubusercontent.com/scala-js/scala-js/v1.7.1/library/src/main/scala/scala/scalajs/runtime/AnonFunctions.scala:27:38)
  at cats.effect.tracing.Tracing$.calculateTracingEvent(https://raw.githubusercontent.com/typelevel/cats-effect/f40765f2421487f83537dfda8829266970d6997c/core/js/src/main/scala/cats/effect/tracing/TracingPlatform.scala:42:26)
  at cats.effect.IO$Delay.flatMap(https://raw.githubusercontent.com/typelevel/cats-effect/f40765f2421487f83537dfda8829266970d6997c/core/shared/src/main/scala/cats/effect/IO.scala:378:54)
  at null.$f_Lcats_effect_kernel_Sync__defer__F0__O(https://raw.githubusercontent.com/typelevel/cats-effect/f40765f2421487f83537dfda8829266970d6997c/core/shared/src/main/scala/cats/effect/IO.scala:1544:17)
* at cats.effect.kernel.Deferred$AsyncDeferred.get(https://raw.githubusercontent.com/typelevel/cats-effect/f40765f2421487f83537dfda8829266970d6997c/core/shared/src/main/scala/cats/effect/IO.scala:1457:57)
+ at null.<anonymous>(https://raw.githubusercontent.com/typelevel/cats-effect/f40765f2421487f83537dfda8829266970d6997c/kernel/shared/src/main/scala/cats/effect/kernel/GenConcurrent.scala:119:23)
  at scala.scalajs.runtime.AnonFunction1.apply(https://raw.githubusercontent.com/scala-js/scala-js/v1.7.1/library/src/main/scala/scala/scalajs/runtime/AnonFunctions.scala:27:38)
  at null.$p_Lcats_effect_IOFiber__succeeded__O__I__Lcats_effect_IO(https://raw.githubusercontent.com/typelevel/cats-effect/f40765f2421487f83537dfda8829266970d6997c/core/shared/src/main/scala/cats/effect/IOFiber.scala:1128:14)
  at null.$p_Lcats_effect_IOFiber__runLoop__Lcats_effect_IO__I__I__V(https://raw.githubusercontent.com/typelevel/cats-effect/f40765f2421487f83537dfda8829266970d6997c/core/shared/src/main/scala/cats/effect/IOFiber.scala:849:28)
```

A couple observations:
1. Lots of inlining happening. For example `AsyncDeferred.get` points to a line in `IO.scala`!
2. Not a single one of those frames points to userland code, so they are all filtered. I think the one indicated `+` is the one we _should_ be keeping, but there's no way to know that.

**Comment by armanbilge** [comment]
So ... despite my best intentions, I think this came full circle 🙃 pretty sure the code I have here is essentially the same as what we already had, give-or-take a couple tiny things. But I think the underlying limitations have been illuminated. Also, I think it's pretty confirmed that Node.js/Chrome do basically the same thing as we always suspected and Firefox is the outlier.

Let me add some comments for next time I forget and decide to go through this ordeal again 😆

**Comment by armanbilge** [comment]
This is what the fibonacci example looks like on Firefox without sourcemaps.
```
The 20th fibonacci number is 17711
java.lang.Throwable: Boom!
  at $c_jl_Throwable(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:13260:5)
  at Example$.program(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15779:41)
  at delay @ Example$.program(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15776)
  at ifM @ Example$.program(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15784)
  at delay @ Example$.program(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15769)
  at flatMap @ Example$.program(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15769)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at delay @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.fib(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15750)
  at flatMap @ Example$.program(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:15758)
  at main @ Example.main(file:///workspace/cats-effect/example/js/target/scala-2.13/cats-effect-example-fastopt/main.js:978)
```

**Review by djspiewak — APPROVED** [review]
Will merge since a side-grade is still worthwhile :-)

**Pull request merged**